### DISCUSSION TITLE
Error when finetune with "wide_multivariate" data type

### DISCUSSION THREAD

**Discussion by HenryLee829** [SalesforceAIResearch/uni2ts#95] [Q&A]
When I use 
`python -m uni2ts.data.builder.simple train Datasets/data.csv --date_offset '2015-12-27 07:36:00'`
 to generate a wide .csv file to an arrow file, then use
`python -m cli.train -cp conf/finetune run_name=train model=moirai_1.0_R_small data=train val_data=train`
to finetune the model, it works out well.
However when I want to try to change the data type into wide_multivariate, I use
`python -m uni2ts.data.builder.simple train Datasets/data.csv --date_offset '2015-12-27 07:36:00' --dataset_type wide_multivariate`
and using the same command to finetune the model, there's an error:
AssertionError: Sample length must be less than or equal to max_length (512)
I don't know how to fix this. Could you give me some advice? Thanks a lot.

**Comment by gorold** [comment]
I'm guessing your time series are too short

**Reply by HenryLee829** [reply]
Thanks a lot for answering! Actually the time series I used to generate arrow file has the shape of (475200, 51), so I think short might not be the problem. I also tried other datasets like the ETTH dataset and use it to generate an "wide_multivariate" arrow file then use it to finetune the model and it has the same error.

I found the assert line in "loader.py", PackCollate class:
`        assert all(
            [len(sample[self.target_field]) <= self.max_length for sample in batch]
        ), f"Sample length must be less than or equal to max_length ({self.max_length})"`

I printed `len(sample[self.target_field])` and I found out that all sample length are multiples of the number of variables and is much larger than the set max_length(512). I think this is due to the method in your paper "flattening a multivariate time series to consider all variates as a single sequence".  But when I set max_length to a larger number, it takes more memory to run. I want to know if there's something I'm doing wrong or there's anything else I can do.
 
 Also, I tried a dataset with two variables and no errors were reported. I used the"moirai_1.0_R_small" model. But the finetuning process took only 5 epochs and stopped to improve. The result is not very good. The following three figures show the finetuning process, predictions before finetuning and predictions after finetuning respectively. I want to know if there's something I'm doing wrong.
![fine-tune](https://github.com/user-attachments/assets/9742427f-becb-41cd-b903-f36941fd2f46)
![before finetune](https://github.com/user-attachments/assets/1dd8a18a-5f25-4960-89d1-52e1893b0259)
![result](https://github.com/user-attachments/assets/35229d53-b939-4348-84f2-1db19a1441f2)

**Reply by lukc12138** [reply]
> Thanks a lot for answering! Actually the time series I used to generate arrow file has the shape of (475200, 51), so I think short might not be the problem. I also tried other datasets like the ETTH dataset and use it to generate an "wide_multivariate" arrow file then use it to finetune the model and it has the same error.
> 
> I found the assert line in "loader.py", PackCollate class: ` assert all( [len(sample[self.target_field]) <= self.max_length for sample in batch] ), f"Sample length must be less than or equal to max_length ({self.max_length})"`
> 
> I printed `len(sample[self.target_field])` and I found out that all sample length are multiples of the number of variables and is much larger than the set max_length(512). I think this is due to the method in your paper "flattening a multivariate time series to consider all variates as a single sequence". But when I set max_length to a larger number, it takes more memory to run. I want to know if there's something I'm doing wrong or there's anything else I can do.
> 
> Also, I tried a dataset with two variables and no errors were reported. I used the"moirai_1.0_R_small" model. But the finetuning process took only 5 epochs and stopped to improve. The result is not very good. The following three figures show the finetuning process, predictions before finetuning and predictions after finetuning respectively. I want to know if there's something I'm doing wrong. ![fine-tune](https://private-user-images.githubusercontent.com/60307783/354462584-9742427f-becb-41cd-b903-f36941fd2f46.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI4MTgzMjksIm5iZiI6MTcyMjgxODAyOSwicGF0aCI6Ii82MDMwNzc4My8zNTQ0NjI1ODQtOTc0MjQyN2YtYmVjYi00MWNkLWI5MDMtZjM2OTQxZmQyZjQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA1VDAwMzM0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJlYTJhMGMwMDlkZGE0YTk0MTBkMTVjYWRkYzM3YTIyMDZjMDA1MDRhNGI4OGJjYzFhMzhlNDdiOWM2NmFiNjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.I798p8jSJuqfkT7u_7XI6vyipQDk3hxbYubOpvcmKSQ) ![before finetune](https://private-user-images.githubusercontent.com/60307783/354462972-1dd8a18a-5f25-4960-89d1-52e1893b0259.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI4MTgzMjksIm5iZiI6MTcyMjgxODAyOSwicGF0aCI6Ii82MDMwNzc4My8zNTQ0NjI5NzItMWRkOGExOGEtNWYyNS00OTYwLTg5ZDEtNTJlMTg5M2IwMjU5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA1VDAwMzM0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVlNTNlNTE3MWZkMDhiYTQzYTY4NjUwNjk0ZGJkZGY2ODU1Y2Y0NDQ5MDFlZjA5NWNkN2U3ZWJiZGVlYmU2NTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.YYV9RkBEDBRCZPKwwvFXBuvcCAXr4-8zf-SG5mWigMg) ![result](https://private-user-images.githubusercontent.com/60307783/354462630-35229d53-b939-4348-84f2-1db19a1441f2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI4MTgzMjksIm5iZiI6MTcyMjgxODAyOSwicGF0aCI6Ii82MDMwNzc4My8zNTQ0NjI2MzAtMzUyMjlkNTMtYjkzOS00MzQ4LTg0ZjItMWRiMTlhMTQ0MWYyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA1VDAwMzM0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTliNGI5MzgyZTVlMzAwMjI2NjU1ZmUyODBjZjQ3ZjFkZWRlMGMyYjRiNmRlMGMzOTU3MDc3ZDViNDVkMGZjOGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.NrYXdp7M3L_n8193csu09CuYs_40i70WfqC9QxqRqf0)

I got the same problem as you mentioned in paragragh 4. Did you solve it?

**Reply by HenryLee829** [reply]
Sorry, I haven't solved it yet. Have you experienced the "max_length" problem I mentioned above(If you are using "wide_multivariate" type)?

**Comment by USTCdyf** [comment]
Hello, may I ask you a question? What’s the difference between the dataset_type options: 'wide' and 'wide_multivariate'? If I use 'wide,' is it still possible to perform multivariate time series prediction? For instance, my dataset contains 7 channels—can I predict all of them simultaneously using 'wide' instead of 'wide_multivariate'? I’m asking because when I processed the data using 'wide_multivariate,' it resulted in only one sample. I don't know whether it is correct

**Reply by HenryLee829** [reply]
Based on my personal understanding, 'wide' is used for handling multiple univariate data sets where there is no correlation between the variables, while 'wide_multivariate' is used for handling a single multivariate data set where there is some correlation between the variables. If your 7 variables are independent, then you can use 'wide' to process them. If you use 'wide_multivariate', it will indeed generate only one sample, but it will include 7 variables. The above is my personal understanding. Since I haven't obtained the desired results either, please correct me if my understanding is incorrect.

**Reply by USTCdyf** [reply]
Sorry for the late reply. My understanding is the same as yours. When I was debugging, I found that using "wide_multivariate" gives a num_samples of 1, while using "wide" gives 7. However, I wanted to ask if you were able to get results for all channels when using "wide" for prediction? For example, in my case, there are 7 channels to predict, but when I use "wide," it seems like I'm only getting the results for the first channel.
